src/Moq/Protected/ProtectedMock.cs

```diff
@@ -443,6 +666,16 @@ private static void ThrowIfVoidMethod(MethodInfo method)
 			}
 		}
 
+		private static object[] CastValueForIndexer(object value,string propertyName)
+		{
+			if (value is object[])
+			{
+				return value as object[];
+			}
+
+			throw new ArgumentException($"{propertyName} is an indexer but value is not an object array");
```

Comment:
This feels wrong.

Why do we even need this method, shouldn't it be replaced with an assertion?

Let me trace this back to one of the callers, `GetPropertyIncludingIndexer`:

![image](https://user-images.githubusercontent.com/104481/130349556-30601cc5-d0ce-471a-b394-9b8914f28098.png)

I find it strange to use the same parameter for a property's `value` as for an indexer's `indexerKeys`. Those are different things. Why are they passed via the same parameter?

This may need some refactoring to make it more self-explanatory just what exactly is going on.

Reply:
@stakx 
The callers of GetPropertyIncludingIndexer do not know if it is a property or an indexer.  The value parameter can be an object or an object[].  It is only used if the propertyName is an indexer and the indexer is overloaded.  Of course if it is an indexer then it should be an object[].  We need to cast and the CastValueForIndexer allows for an appropriate error message.

Note that the GetPropertyIncludingIndexer can be called for methods too under certain circumstances.

The parameter name 'value' can be changed if you would like.  Suggestions ?

can be a method